ray-operator/controllers/raycluster_controller.go

```diff
@@ -242,7 +311,56 @@ func (r *RayClusterReconciler) reconcilePods(instance *rayiov1alpha1.RayCluster)
 				return err
 			}
 		}
+	} else {
+		// head node is exactly 1, but we need to check if it has been changed
+		res := r.needRemoveOldPod(instance.Spec.HeadGroupSpec.Template, headPods.Items[0])
+		if res {
+			log.Info(fmt.Sprintf("need to delete old head pod %s", headPods.Items[0].Name))
+			if err := r.Delete(context.TODO(), &headPods.Items[0]); err != nil {
+				return err
+			}
+			// delete all worker pods
+			workerPods := corev1.PodList{}
+			filterLabels = client.MatchingLabels{common.RayClusterLabelKey: instance.Name}
+			if err := r.List(context.TODO(), &workerPods, client.InNamespace(instance.Namespace), filterLabels); err != nil {
+				return err
+			}
+			for _, aPod := range workerPods.Items {
+				log.Info(fmt.Sprintf("need to delete old worker pod %s", aPod.Name))
+				if (aPod.Status.Phase == v1.PodRunning || aPod.Status.Phase == v1.PodPending) && aPod.ObjectMeta.DeletionTimestamp == nil {
+					if err := r.Delete(context.TODO(), &aPod); err != nil {
+						log.Info(fmt.Sprintf("error deleting worker pod %s", aPod.Name))
+						return err
+					}
+				}
+			}
+			return nil
+		}
+	}
+
+	// check if WorkerGroupSpecs has been changed and we need to kill worker pods
+	for _, worker := range instance.Spec.WorkerGroupSpecs {
+		workerPods := corev1.PodList{}
+		filterLabels = client.MatchingLabels{common.RayClusterLabelKey: instance.Name, common.RayNodeGroupLabelKey: worker.GroupName}
+		if err := r.List(context.TODO(), &workerPods, client.InNamespace(instance.Namespace), filterLabels); err != nil {
+			return err
+		}
+		updatedWorkerPods := false
+		for _, item := range workerPods.Items {
+			if r.needRemoveOldPod(worker.Template, item) {
```

Comment:
we actually has the worker group template. In that case, the template would be same. Do we need to compare pod with template one by one?

Reply:
In this case, the group template length would be 1. So anyway we will not have extra overhead here, right?

Reply:
Yes. the workingGroupTemplate would be 1. I mean the pods created from the template should be same. Do we want to compare each of them with the template or just randomly pick one? I don't have strong opinion on this because technically, the pods could be different if there're some manual modification.